src/ray/core_worker/fiber.h

```diff
@@ -0,0 +1,121 @@
+#ifndef RAY_CORE_WORKER_FIBER_H
+#define RAY_CORE_WORKER_FIBER_H
+
+#include <ray/util/logging.h>
+#include <boost/fiber/all.hpp>
+namespace ray {
+
+/// Used by async actor mode. The fiber event will be used
+/// from python to switch control among different coroutines.
+/// Taken from boost::fiber examples
+/// https://github.com/boostorg/fiber/blob/7be4f860e733a92d2fa80a848dd110df009a20e1/examples/wait_stuff.cpp#L115-L142
+class FiberEvent {
+ public:
+  // Block the fiber until the event is notified.
+  void Wait() {
+    std::unique_lock<boost::fibers::mutex> lock(mutex_);
+    cond_.wait(lock, [this]() { return ready_; });
+  }
+
+  // Notify the event and unblock all waiters.
+  void Notify() {
+    {
+      std::unique_lock<boost::fibers::mutex> lock(mutex_);
+      ready_ = true;
+    }
+    cond_.notify_one();
+  }
+
+ private:
+  boost::fibers::condition_variable cond_;
+  boost::fibers::mutex mutex_;
+  bool ready_ = false;
+};
+
+/// Used by async actor mode. The FiberRateLimiter is a barrier that
+/// allows at most num fibers running at once. It implements the
+/// semaphore data structure.
+class FiberRateLimiter {
+ public:
+  FiberRateLimiter(int num) : num_(num) {}
+
+  // Enter the semaphore. Wait fo the value to be > 0 and decrement the value.
+  void Acquire() {
+    std::unique_lock<boost::fibers::mutex> lock(mutex_);
+    cond_.wait(lock, [this]() { return num_ > 0; });
+    num_ -= 1;
+  }
+
+  // Exit the semaphore. Increment the value and notify other waiter.
+  void Release() {
+    {
+      std::unique_lock<boost::fibers::mutex> lock(mutex_);
+      num_ += 1;
+    }
+    // TODO(simon): This not does guarantee to wake up the first queued fiber.
+    // This could be a problem for certain workloads because there is no guarantee
+    // on task ordering .
+    cond_.notify_one();
+  }
+
+ private:
+  boost::fibers::condition_variable cond_;
+  boost::fibers::mutex mutex_;
+  int num_ = 1;
+};
+
+using FiberChannel = boost::fibers::unbuffered_channel<std::function<void()>>;
+
+class FiberState {
+ public:
+  FiberState(int max_concurrency) : rate_limiter_(max_concurrency) {
+    fiber_runner_thread_ = std::thread([&]() {
+      while (!channel_.is_closed()) {
+        std::function<void()> func;
+        auto op_status = channel_.pop(func);
+        if (op_status == boost::fibers::channel_op_status::success) {
+          boost::fibers::fiber(boost::fibers::launch::dispatch, func).detach();
+        } else {  // Errored
+          RAY_LOG(ERROR) << "Fiber worker popped errrored. Exiting";
```

Comment:
typos and this is a really bad error message to surface to users. maybe add some more context?